File: app/src/main/java/com/think/runex/java/Models/WorkoutInfo.java
package com.think.runex.java.Models;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.android.gms.maps.model.LatLng;
import com.think.runex.java.App.Configs;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import static com.think.runex.util.ConstantsKt.DISPLAY_DATE_TIME_FORMAT_THREE_LETTERS_DATE_MONTH;


public class WorkoutInfo implements Parcelable {
    /**
     * id : 5dc02b9c874a693a72ef612e
     * distance : 0.01
     * img_url :
     * caption :
     * activity_date : 2019-11-04T13:46:04.662Z
     * created_at : 2019-11-04T13:46:04.741Z
     * updated_at : 2019-11-04T13:46:04.741Z
     */

    private String id;
    private String activity_type;
    private String app;
    private String ref_id;
    private double calory;
    private String caption;
    private double distance;
    private double pace;
    private int duration;
    private String time_string;
    private String start_date;
    private String end_date;
    private String workout_date;
    private int net_elevation_gain;
    private boolean is_sync;
    private List<RealmPointObject> locations;

    public WorkoutInfo() {
    }


    protected WorkoutInfo(Parcel in) {
        id = in.readString();
        activity_type = in.readString();
        app = in.readString();
        ref_id = in.readString();
        calory = in.readDouble();
        caption = in.readString();
        distance = in.readDouble();
        pace = in.readDouble();
        duration = in.readInt();
        time_string = in.readString();
        start_date = in.readString();
        end_date = in.readString();
        workout_date = in.readString();
        net_elevation_gain = in.readInt();
        is_sync = in.readByte() != 0;
        locations = in.createTypedArrayList(RealmPointObject.CREATOR);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(id);
        dest.writeString(activity_type);
        dest.writeString(app);
        dest.writeString(ref_id);
        dest.writeDouble(calory);
        dest.writeString(caption);
        dest.writeDouble(distance);
        dest.writeDouble(pace);
        dest.writeInt(duration);
        dest.writeString(time_string);
        dest.writeString(start_date);
        dest.writeString(end_date);
        dest.writeString(workout_date);
        dest.writeInt(net_elevation_gain);
        dest.writeByte((byte) (is_sync ? 1 : 0));
        dest.writeTypedList(locations);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public static final Creator<WorkoutInfo> CREATOR = new Creator<WorkoutInfo>() {
        @Override
        public WorkoutInfo createFromParcel(Parcel in) {
            return new WorkoutInfo(in);
        }

        @Override
        public WorkoutInfo[] newArray(int size) {
            return new WorkoutInfo[size];
        }
    };

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getActivity_type() {
        return activity_type;
    }

    public void setActivity_type(String activity_type) {
        this.activity_type = activity_type;
    }

    public String getApp() {
        return app;
    }

    public void setApp(String app) {
        this.app = app;
    }

    public String getRef_id() {
        return ref_id;
    }

    public void setRef_id(String ref_id) {
        this.ref_id = ref_id;
    }

    public double getCalory() {
        return calory;
    }

    public void setCalory(double calory) {
        this.calory = calory;
    }

    public String getCaption() {
        return caption;
    }

    public void setCaption(String caption) {
        this.caption = caption;
    }

    public double getDistance() {
        return distance;
    }

    public void setDistance(double distance) {
        this.distance = distance;
    }

    public double getPace() {
        return pace;
    }

    public void setPace(double pace) {
        this.pace = pace;
    }

    public int getDuration() {
        return duration;
    }

    public void setDuration(int duration) {
        this.duration = duration;
    }

    public String getTime_string() {
        return time_string;
    }

    public void setTime_string(String time_string) {
        this.time_string = time_string;
    }

    public String getStart_date() {
        return start_date;
    }

    public void setStart_date(String start_date) {
        this.start_date = start_date;
    }

    public String getEnd_date() {
        return end_date;
    }

    public void setEnd_date(String end_date) {
        this.end_date = end_date;
    }

    public String getWorkout_date() {
        return workout_date;
    }

    public String getWorkoutDate() {
        if (workout_date == null) {
            return "";
        }
        String changedDateTime = workout_date;
        try {
            SimpleDateFormat sdf = new SimpleDateFormat(Configs.SERVER_DATE_TIME_FORMAT, Locale.getDefault());
            Date date = sdf.parse(workout_date);
            if (date != null) {
                changedDateTime = new SimpleDateFormat(DISPLAY_DATE_TIME_FORMAT_THREE_LETTERS_DATE_MONTH, Locale.getDefault()).format(date);
            }
        } catch (Throwable error) {
            try {
                error.printStackTrace();
                SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SS'Z'", Locale.getDefault());
                Date date = null;

                date = sdf.parse(workout_date);

                if (date != null) {
                    changedDateTime = new SimpleDateFormat(DISPLAY_DATE_TIME_FORMAT_THREE_LETTERS_DATE_MONTH, Locale.getDefault()).format(date);
                }
            } catch (Throwable e) {
                e.printStackTrace();
            }
        }
        return changedDateTime;
    }

    public void setWorkout_date(String workout_date) {
        this.workout_date = workout_date;
    }

    public int getNet_elevation_gain() {
        return net_elevation_gain;
    }

    public void setNet_elevation_gain(int net_elevation_gain) {
        this.net_elevation_gain = net_elevation_gain;
    }

    public boolean isIs_sync() {
        return is_sync;
    }

    public void setIs_sync(boolean is_sync) {
        this.is_sync = is_sync;
    }

    public List<RealmPointObject> getLocations() {
        return locations;
    }

    public void setLocations(List<RealmPointObject> locations) {
        this.locations = locations;
    }

    public List<LatLng> getLatLngList() {
        ArrayList<LatLng> latLngList = new ArrayList<>();
        for (RealmPointObject point : locations) {
            latLngList.add(point.toLatLng());
        }
        return latLngList;
    }

    public RealmRecorderObject toRealmRecorderObject() {
        RealmRecorderObject object = new RealmRecorderObject();

        object.setWorkout_date(workout_date);
        object.setDistanceKm(distance);
        object.setDurationMillis((duration * 1000));
        object.setPaceMillis((long) ((pace * 60) * 1000));
        object.setDisplayRecordAsTime(time_string);
        //object.setDisplayPaceAsTime();
        object.setCalories(calory);

        return object;
    }
}

File: app/src/main/java/com/think/runex/feature/workout/data/WorkoutInfo.kt
package com.think.runex.feature.workout.data

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import androidx.annotation.DrawableRes
import com.google.gson.annotations.SerializedName
import com.jozzee.android.core.datetime.dateTimeFormat
import com.jozzee.android.core.datetime.toTimeMillis
import com.think.runex.R
import com.think.runex.config.DISPLAY_DATE_FORMAT
import com.think.runex.util.extension.displayFormat
import com.think.runex.util.extension.timeDisplayFormat
import com.think.runex.config.DISPLAY_DATE_TIME_FORMAT_THREE_LETTERS_DATE_MONTH
import com.think.runex.config.SERVER_DATE_TIME_FORMAT
import com.think.runex.config.SERVER_DATE_TIME_FORMAT_2
import java.text.SimpleDateFormat
import java.util.*

data class WorkoutInfo(
    @SerializedName("activity_type") var activityType: String? = null,
    @SerializedName("app") var app: String? = null,
    @SerializedName("calory") var calories: Double? = 0.0,
    @SerializedName("caption") var caption: String? = null,
    @SerializedName("distance") var distanceKilometers: Float? = 0f,
    @SerializedName("duration") var durationSecond: Long? = 0,
    @SerializedName("end_date") var endDate: String? = null,
    @SerializedName("id") var id: String? = null,
    @SerializedName("is_sync") var isSync: Boolean? = false,
    @SerializedName("net_elevation_gain") var netElevationGain: Double? = 0.0,
    @SerializedName("locations") var locations: List<WorkingOutLocation>? = null,
    @SerializedName("pace") var durationMinutePerKilometer: Double? = 0.0,
    @SerializedName("ref_id") var refId: String? = null,
    @SerializedName("start_date") var startDate: String? = null,
    @SerializedName("time_string") var timeDisplay: String? = null,
    @SerializedName("workout_date") var workoutDate: String? = null,
    @SerializedName("loc_url") var locationsUrl: String? = null
) : Parcelable {

    companion object CREATOR : Parcelable.Creator<WorkoutInfo> {
        override fun createFromParcel(parcel: Parcel): WorkoutInfo {
            return WorkoutInfo(parcel)
        }

        override fun newArray(size: Int): Array<WorkoutInfo?> {
            return arrayOfNulls(size)
        }
    }

    constructor(parcel: Parcel) : this(
        parcel.readString(),
        parcel.readString(),
        parcel.readValue(Double::class.java.classLoader) as? Double,
        parcel.readString(),
        parcel.readValue(Float::class.java.classLoader) as? Float,
        parcel.readValue(Long::class.java.classLoader) as? Long,
        parcel.readString(),
        parcel.readString(),
        parcel.readValue(Boolean::class.java.classLoader) as? Boolean,
        parcel.readValue(Double::class.java.classLoader) as? Double,
        parcel.createTypedArrayList(WorkingOutLocation),
        parcel.readValue(Double::class.java.classLoader) as? Double,
        parcel.readString(),
        parcel.readString(),
        parcel.readString(),
        parcel.readString(),
        parcel.readString()
    )

    constructor(record: WorkingOutRecord?, locations: List<WorkingOutLocation>?) : this(
        activityType = record?.type ?: WorkoutType.RUNNING,
        app = "RUNEX",
        calories = record?.getCalories() ?: 0.0,
        caption = "",
        distanceKilometers = record?.getDistancesKilometers() ?: 0f,
        durationSecond = record?.getDurationSecond() ?: 0,
        endDate = record?.stop?.dateTimeFormat(SERVER_DATE_TIME_FORMAT) ?: "",
        id = "",
        isSync = false,
        netElevationGain = 0.0,
        locations = locations ?: emptyList(),
        durationMinutePerKilometer = record?.getDurationMinutePerKilometer() ?: 0.0,
        refId = "",
        startDate = record?.start?.dateTimeFormat(SERVER_DATE_TIME_FORMAT) ?: "",
        workoutDate = record?.start?.dateTimeFormat(SERVER_DATE_TIME_FORMAT) ?: "",
        timeDisplay = record?.duration?.timeDisplayFormat() ?: ""
    )


    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(activityType)
        parcel.writeString(app)
        parcel.writeValue(calories)
        parcel.writeString(caption)
        parcel.writeValue(distanceKilometers)
        parcel.writeValue(durationSecond)
        parcel.writeString(endDate)
        parcel.writeString(id)
        parcel.writeValue(isSync)
        parcel.writeValue(netElevationGain)
        parcel.writeTypedList(locations)
        parcel.writeValue(durationMinutePerKilometer)
        parcel.writeString(refId)
        parcel.writeString(startDate)
        parcel.writeString(timeDisplay)
        parcel.writeString(workoutDate)
        parcel.writeString(locationsUrl)
    }

    override fun describeContents(): Int {
        return 0
    }

    fun getDisplayData(): WorkingOutDisplayData {

        val displayData = WorkingOutDisplayData()

        //Update distance
        displayData.distances = distanceKilometers?.displayFormat(awaysShowDecimal = true) ?: "0.00"

        //Update duration
        displayData.duration = timeDisplay
            ?: "00:00:00"//((durationSecond ?: 0L) * 1000).timeDisplayFormat()

        //Update duration per kilometer
        if (distanceKilometers ?: 0f > 0) {
            val millisPerKilometer: Long = (((durationSecond ?: 0) / (distanceKilometers
                ?: 0f)) * 1000).toLong()
            //Check durations millisecond per kilometer more than 1 hour (3600000 millisecond)
            when (millisPerKilometer < 3600000) {
                true -> displayData.durationPerKilometer = millisPerKilometer.timeDisplayFormat().let {
                    if (it.length >= 3) {
                        return@let it.substring(3, it.length)
                    }
                    return@let it
                }
                false -> {
                    displayData.durationPerKilometer = millisPerKilometer.timeDisplayFormat()
                    displayData.durationPerKilometerUnit = "(hr/km)"
                }
            }
        }

        //Update calories
        displayData.calories = calories?.displayFormat() ?: "0"

        return displayData
    }

    fun getWorkoutDateTime(displayPattern: String = DISPLAY_DATE_TIME_FORMAT_THREE_LETTERS_DATE_MONTH): String {
        var text = ""
        try {
            val sdf = SimpleDateFormat(SERVER_DATE_TIME_FORMAT, Locale.getDefault())
            sdf.timeZone = TimeZone.getDefault()
            sdf.parse(workoutDate ?: "")?.also { date ->
                text = SimpleDateFormat(displayPattern, Locale.getDefault()).format(date)
            }
        } catch (e: Throwable) {
            try {
                val sdf = SimpleDateFormat(SERVER_DATE_TIME_FORMAT_2, Locale.getDefault())
                sdf.timeZone = TimeZone.getDefault()
                sdf.parse(workoutDate ?: "")?.also { date ->
                    text = SimpleDateFormat(displayPattern, Locale.getDefault()).format(date)
                }
            } catch (e: Throwable) {
                try {
                    val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
                    sdf.timeZone = TimeZone.getDefault()
                    sdf.parse(workoutDate ?: "")?.also { date ->
                        text = SimpleDateFormat(displayPattern, Locale.getDefault()).format(date)
                    }
                } catch (e: Throwable) {
                }
            }
        }
        return text
    }

    fun getWorkoutDateTimeMillis(): Long {
        var time: Long = 0
        try {
            time = SimpleDateFormat(SERVER_DATE_TIME_FORMAT, Locale.getDefault())
                .parse(workoutDate ?: "")?.time ?: 0
        } catch (e: Throwable) {
            try {
                time = SimpleDateFormat(SERVER_DATE_TIME_FORMAT_2, Locale.getDefault())
                    .parse(workoutDate ?: "")?.time ?: 0
            } catch (e: Throwable) {
                try {
                    time = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
                        .parse(workoutDate ?: "")?.time ?: 0
                } catch (e: Throwable) {

                }
            }
        }
        return time
    }

    fun getWorkoutDateTimeForImageName(): String {
        return workoutDate?.dateTimeFormat(SERVER_DATE_TIME_FORMAT, "yyyyMMddHHmm") ?: ""
    }

    fun getDistances(context: Context): String {
        return "${distanceKilometers?.displayFormat(awaysShowDecimal = true) ?: ""} ${context.getString(R.string.km)}"
    }

    @DrawableRes
    fun getPartnerIcon(): Int = when {
        app?.contains(Partner.GARMIN, true) == true -> R.mipmap.ic_logo_garmin
        app?.contains(Partner.SUUNTO, true) == true -> R.mipmap.ic_logo_suunto
        app?.contains(Partner.STRAVA, true) == true -> R.mipmap.ic_logo_strava
        app?.contains(Partner.FITBIT, true) == true -> R.mipmap.ic_logo_fitbit
        else -> R.drawable.ic_running
    }
}